onnxruntime/contrib_ops/cpu/bert/attention_helper.h

```diff
@@ -20,33 +20,34 @@ template <typename T>
 void ComputeSmoothSoftmaxInplace(T* score, int N, int D, ThreadPool* tp) {
   ThreadPool::TryParallelFor(tp, N, D * 2.0, [&](std::ptrdiff_t begin, std::ptrdiff_t end) {
     for (std::ptrdiff_t j = begin; j != end; ++j) {
-      float* x = reinterpret_cast<T*>(score) + j * D;
-      float* y = x;
+      T* x = reinterpret_cast<T*>(score) + j * D;
+      T* y = x;
 
       float max = -std::numeric_limits<float>::infinity();
       for (int i = 0; i < D; i++) {
-        if (max < x[i])
-          max = x[i];
+        float x_i = static_cast<float>(x[i]);
+        if (max < x_i)
+          max = x_i;
       }
 
       if (max < 0.0f) {
         max = 0.0f;
       }
 
       for (int i = 0; i < D; i++) {
-        y[i] = expf(x[i] - max);
+        y[i] = static_cast<T>(expf(static_cast<float>(x[i]) - max));
```

Comment:
when T is float16, it will overflow easily. 
I think we cannot use softmax inplace for float16. A float buffer is needed.

Reply:
should the expf(x[i] - max) belong to (0, 1]?

Reply:
Right. It will not overflow.
I think we need to keep intermediate data as float/double to keep the accuracy. Every time data casted from float to half will cause accuracy loss. The loss is accumulated when we compute the sum below.